### DISCUSSION TITLE
Compression before encryption seems to weakening the cypher

### DISCUSSION THREAD

**Discussion by evgmik** [seaweedfs/seaweedfs#6022] [Ideas]
It was my own PR but do changes in 0da9e40 decrease security of encryption after compression?

Here is the problem: gzip format has standard 10 bytes header (see https://datatracker.ietf.org/doc/html/rfc1952). 

The first two bytes of each are always "1f 8b" (gzip file format magic ID) the next "08"  represents compression method (where 08 stands for gzip),  and there are
4 bytes for data stamp, which are guessable, there is one more byte which is guessable, the one which encodes OS with countable number of choices. Two bytes which encode regular and special flags are limited in options too. 

In the essence, the first 10 bytes of the data are super weak. I am not an expert in encoding, but I presume if the attacker can guess 10 bytes of the file at the beginning, this probably decrease the strength of the cypher key (in the worst case by 10 bytes).

Our current key is 32 bytes, so there are still a lot of bytes to guess for attacker but the  amount of needed work is decreased.

There are several ways to get out of this:
1. I am wrong in my assumptions of the key weakening
2. reverse the patch
3. make compression optional (at the user risk) if encryption is on
4. increase key length (probably the easiest but it cost CPU and data structure rearrangement).
5. Double encryption (would it help?)
6. strip away gzip headers before encryption and store them in the filer or in a different needle

The last option seems to be a good idea:
1st, even without encryption, gzip `extra` header might contain a file name (probably not a case for SeaweedFS)
2nd, it might allow to use other compressors, not only gzip (which is not the best by modern standards). I.e. SeaweedFS can read de compressor based on the headers, and we will move away from hard coded gzip.

Any thoughts?

**Comment by dmi3x3** [comment]
You wrote to me in another [topic](https://github.com/seaweedfs/seaweedfs/discussions/5524), there were mentioned two links about hacking compressed data,  [CRIME](https://en.wikipedia.org/wiki/CRIME?__ya_mt_enable_static_translations=1) 
 [BREACH](https://en.wikipedia.org/wiki/BREACH?__ya_mt_enable_static_translations=1).
In general, any formatted structured data is easier to hack (the format is already a kind of hint), but you need the ability to get the same file several times with different options.

1. Theoretically goes down.
2. Do not cancel:) - until recently there was no compression, now it will be, who needs increased protection, will not use compression.
3. Compression is still optional.
4. If possible, optional.
5. Double encryption does not make sense, the encrypted file will not be compressed.
6. It looks good, but will the reliability of file storage suffer?

Thanks for the [0da9e40](https://github.com/seaweedfs/seaweedfs/commit/0da9e40abb9628831f47656b4b62a9d72003e9bf)

**Reply by evgmik** [reply]
Thanks for the pointers. Quite a fascinating read, I did not know that https could be breached without brute force.

But the good news is that the seaweedfs encryption cannot be compomised by CRIME/BREACH type of attack. They both require reuse of encryption key and somesort of response to the user supplied data.

Seaweedfs uses single random key per chunk. So there is no reuse and it does not change its response to a chunk requests. So the data stored at volume servers seems to be safe.

The attack which I was worried is in the category of the partially known-plaintext attack. Apparently modern cyphers (such as AES used by seeaweedfs) are [fairly immune to it](https://crypto.stackexchange.com/questions/45110/encrypting-a-partially-known-plain-text). We are safe on this front this as well.

Thus, I conclude that enabling encryption after compression does not compromise seaweedfs storage. It is still safe to keep such data on potentially compromised locations. The adversary would spent similar amount of time cracking such storage format as with encryption without comression.
